File: 1268_iterator_combination/combination.py
class CombinationIterator:
    def __init__(self, characters: str, combinationLength: int):
        """
        Time: O(2**n * n)
        Space: O(k * C**k)
        """
        self.combinations = []
        n, k = len(characters), combinationLength
        for bitmask in range(1 << n):
            if bin(bitmask).count('1') == k:
                curr = [characters[j]
                        for j in range(n) if bitmask & (1 << n - j - 1)]
                self.combinations.append(''.join(curr))

    def next(self) -> str:
        """
        Time: O(1)
        """
        return self.combinations.pop()

    def hasNext(self) -> bool:
        return self.combinations


# Your CombinationIterator object will be instantiated and called as such:
# obj = CombinationIterator(characters, combinationLength)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: 1268_iterator_combination/combination.java
class CombinationIterator {
    public Deque<String> combinations = new ArrayDeque<String>();

    // Time: O(2**n * n)
    // Space: O(k * C**k)
    public CombinationIterator(String characters, int combinationLength) {
        int n = characters.length();
        int k = combinationLength;

        for (int bitmask = 0; bitmask < 1 << n; bitmask++) {
            if (Integer.bitCount(bitmask) == k) {
                StringBuilder curr = new StringBuilder();
                for (int j = 0; j < n; j++) {
                    if ((bitmask & (1 << n - j - 1)) != 0)
                        curr.append(characters.charAt(j));
                }
                combinations.push(curr.toString());
            }
        }
    }

    // Time: O(1)
    public String next() {
        return combinations.pop();
    }

    public boolean hasNext() {
        return (!combinations.isEmpty());
    }
}

/**
 * Your CombinationIterator object will be instantiated and called as such:
 * CombinationIterator obj = new CombinationIterator(characters,
 * combinationLength); String param_1 = obj.next(); boolean param_2 =
 * obj.hasNext();
 */